teps/0047-finallytask-execution-post-timeout.md

```diff
@@ -0,0 +1,290 @@
+| ∙ hello-world-pipeline-run-with-timeout-task2-44tsb | task2     | 8 seconds ago  | 5 seconds  | Succeeded              |
+| ∙ hello-world-pipeline-run-with-timeout-task1-wgcq7 | task1     | 38 seconds ago | 30 seconds | Failed(TaskRunTimeout) |
+|                                                     |           |                |            |                        |
+
+
+The finally task still executes after the task failure.
+
+
+Finally if we reduce the pipelinerun timeout to 10s, our status report shows:
+
+`PipelineRun "hello-world-pipeline-run-with-timeout" failed to finish within "10s" (TaskRun "hello-world-pipeline-run-with-timeout-task1-q7fw4" failed to finish within "30s")`
+
+| NAME                                                | TASK NAME | STARTED       | DURATION   | STATUS                 |
+|-----------------------------------------------------|-----------|---------------|------------|------------------------|
+| ∙ hello-world-pipeline-run-with-timeout-task1-q7fw4 | task1     | 2 minutes ago | 30 seconds | Failed(TaskRunTimeout) |
+|                                                     |           |               |            |                        |
+|                                                     |           |               |            |                        |
+
+
+The pipelinerun timeout take precedence over the task timeout. After 10s the task fails... And the finally task does not get the chance to execute.
+
+
+For this reason, it is currently not possible to rely on Finally tasks for any of the aforementioned use cases.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+Enable the uses cases :
+
+- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
+- Update Pull Request with what happened overall in the pipeline (pipeline level)
+- Report Test Results at the end of the test pipeline (Notifications Scenario)
+
+When a pipelinerun times out.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Enable finally task to run when a pipeline times out.
+
+We propose to have a new flag ```timeouts``` which would be a dictionary of a set of timeouts. We propose adding ```tasks``` as the first key of that dictionary. 
+
+```yaml
+timeouts:
+  tasks: "0h4m0s"
+  ...
+```
+
+
+`tasks` will define a timeout for the dag tasks. The finally tasks timeout will be `timeout - tasks` with ```timeout```being the current timeout flag and `timeout >= tasks`.
```

Comment:
I feel this is overkill, since we all agree to go in a direction with a new section as a dictionary, we can just isolate this without mixing with the existing `timeout` field in `pipelineRun`. Keep the existing behavior as is. Introduce a new section `timeouts` in `pipelineRun` to read `tasks` timeout, `pipeline` timeout, and `finally` timeout. Users can either specify the traditional timeout or this new timeout dictionary. 

```
kind: PipelineRun
spec:
  timeout: "0h4m0s"
```

or 

```
kind: PipelineRun
spec:
  timeouts:
    pipeline: "0h4m0s"
    tasks: "0h1m0s"
    finally: "0h3m0s"
```

Reply:
My concern with mixing the new field and the traditional usage:

(1) Decoupling both in the future i.e replace `timeout` with `timeouts.pipeline`.
(2) Allowing to specify both `timeout` and `timeouts` right now and then restrict the same usage in the future.

Reply:
added a follow-on work section with the timeout dict proposition., Kept the proposal on the tasksTimeout field.